### ISSUE TITLE
[enhancement] focus of key-verification popup

### ISSUE THREAD

**Issue by r2evans** [dlech/KeeAgent#231]
I have selected "Always require user confirmation...". When the popup appears, it is not Windows-focused, which means my normally keyboard-centric workflow is disrupted in order to select the window; at that point, I can left-key or `Alt-Y` my affirmation.

Is there a way to force making this popup be the Windows focus?

Reproduction:

1. open git-for-windows bash;
2. initiate a connection somewhere, e.g., using `ssh`, `scp`, or `rsync` (not within GfW);
3. popup appears, but is not the focus.

KeePass-2.38; KeeAgent-0.10.1.0; Win10pro; GfW-2.17.0

**Comment by dlech** [comment]
It is supposed to be since #180.

**Comment by r2evans** [comment]
Ok, I'll work on a better reproduction, as it is now working (no changes, no restarts), but I know it has not been focused earlier today. Apologize for an incomplete reproduction.

**Comment by dlech** [comment]
I kind of have a feeling that this is Windows trying to be "smart" and not focus our application even though we requested it. I don't use Windows much but I have noticed that kind of behavior over the years with all applications.

**Comment by r2evans** [comment]
I was thinking the same thing, though I don't have experience programming the Windows-UI side of things. Thanks, I'll report back if I find anything notable.

**Comment by r2evans** [comment]
It's definitely not doing it all of the time: right now, it's always unfocused. It's the same from `cmd.exe` or Git `bash` (in a mintty), using `ssh` or `scp`. I can easily bring it into focus (mouse-click, <kbd>Shift</kbd>-<kbd>Alt</kbd>-<kbd>Tab</kbd>, or twice <kbd>Win</kbd>-<kbd>8</kbd> (or which ever numbered icon KeePass is on the task bar)).

Is there something I can do to try to figure out what is different about this vice your code's intention? If there's a specific debug-able version of the plugin, I'll run that for you.

**Comment by KuttKatrea** [comment]
I have experienced the same problem "randomly" (it almost always works, but sometimes doesn't) but I haven't found the cause of that "sometimes"

I read somewhere that Windows tries not to give focus to background-started programs, maybe certain level of indirection (´git command -> ssh´) causes windows to recognize the window as a background-launched process?

**Comment by KuttKatrea** [comment]
It has just happened to me, when using git in powershell running inside the IntelliJ Terminal.

Changing the *"Esc minimizes to tray..."* option did work for one *git* call on a *cmd* window once, then it stopped working again.

I'll try to keep testing

**Comment by aXe1** [comment]
Similar problems: need to `Alt+Shift+Tab` every time the popup shows. If there are 2 consequent popup (ssh client is trying multiple keys), then only first is unfocused, all following would be ok. No difference from what ssh client call happens: Git Bash, WSL Bash, JetBrains DataGrip, Visual Studio Code, Total Commander.

```
Windows 10 (1809, 17763.316)
KeePass 2.4.1
KeeAgent 0.10.1
```

**Comment by aXe1** [comment]
As a quick and dirty workaround I use [AutoIt](https://www.autoitscript.com/site/) script:
```autoit
While 1
  Local $hwnd = WinWait("KeePass", "has requested to use the SSH key")
  WinActivate($hwnd)
WEnd
```
[Gist](https://gist.github.com/aXe1/183ccc2e95d13d0104d52a5a6af73d46)
[Precompiled x64 version](https://github.com/dlech/KeeAgent/files/3296227/keeagent-autofocus.zip)

**Comment by markkuleinio** [comment]
This is still the number one problem in using KeeAgent. There is no apparent logic when the window opens in focus or not focused.

Here is another workaround (since Shift-Alt-Tab is not my favourite key combinations): Press **Ctrl-Alt-K** (or whatever is your hotkey for "Show KeePass window" in KeePass) and then **Esc** (to minimize KeePass), then the KeeAgent window gets the focus.

**Comment by markkuleinio** [comment]
FYI: Now that I've used **Windows 11** (build 22000.778) for a few days, my initial expecience is that KeeAgent does not suffer from the out-of-focus problem anymore.

**Comment by r2evans** [comment]
I just confirmed that it is still the same. I'm on Win11 (Build 22000.778), KeePass-2.51.1, KeeAgent-0.12.1. The original steps still reproduce the same symptom that the confirmation dialog is not windows-focused when it comes up.

**Comment by markkuleinio** [comment]
I don't know it that matters but I have the key selection dialog enabled, so it is shown first, and then the confirmation dialog. I configured KeePass and KeeAgent identically on Win10 and Win11.

As a side note, on that new Win11 laptop also Evernote Legacy works better (disclaimer: I'm not endorsing Evernote by any means, on the contrary): new note shortcut opens the note in focus. In general both Win10 (Lenovo X1 Carbon Gen 6) and Win11 (X1C Gen 10) laptops have the same apps installed with the same settings, as far as practically possible. No idea what could have affected these focus issues.

**Comment by markkuleinio** [comment]
After upgrading **Windows 11** to **22H2** (build 22621.900) last week I have started to experience the same out-of-focus problems again. Sigh

**Comment by dlech** [comment]
FYI, I just merged a PR that changes the UI thread where the confirmation dialog runs. Not sure if it would make a difference in this issue, but if anyone wants to try the build from https://github.com/dlech/KeeAgent/actions/runs/7603541323 and see...

**Comment by markkuleinio** [comment]
Tested 1.5 days now: 100% accuracy this far, no focus problems! 💪

Windows 11 to 22H2 (build 22621.3007)

**Comment by markkuleinio** [comment]
Looks good based on day 3 with that build!

**Comment by markkuleinio** [comment]
Any plans to make a release with this feature?

**Comment by dlech** [comment]
Thanks for the reminder. I just released v0.13.7 so hopefully we can call this issue closed now.
